pkg/cluster/sync.go

```diff
@@ -845,69 +845,100 @@ func (c *Cluster) syncConnectionPooler(oldSpec,
+			newNeedConnectionPooler = c.needMasterConnectionPoolerWorker(&newSpec.Spec)
+			oldNeedConnectionPooler = c.needMasterConnectionPoolerWorker(&oldSpec.Spec)
+		} else {
+			newNeedConnectionPooler = c.needReplicaConnectionPoolerWorker(&newSpec.Spec)
+			oldNeedConnectionPooler = c.needReplicaConnectionPoolerWorker(&oldSpec.Spec)
+		}
+		if c.ConnectionPooler == nil {
+			c.ConnectionPooler = &ConnectionPoolerObjects{}
+			c.ConnectionPooler.Deployment = make(map[PostgresRole]*appsv1.Deployment)
+			c.ConnectionPooler.Service = make(map[PostgresRole]*v1.Service)
+		}
+
+		if newNeedConnectionPooler {
+			// Try to sync in any case. If we didn't needed connection pooler before,
+			// it means we want to create it. If it was already present, still sync
+			// since it could happen that there is no difference in specs, and all
+			// the resources are remembered, but the deployment was manually deleted
+			// in between
+			c.logger.Debug("syncing connection pooler")
+
+			// in this case also do not forget to install lookup function as for
+			// creating cluster
+			if !oldNeedConnectionPooler || !c.ConnectionPooler.LookupFunction {
+				newConnectionPooler := newSpec.Spec.ConnectionPooler
+
+				specSchema := ""
+				specUser := ""
+
+				if newConnectionPooler != nil {
+					specSchema = newConnectionPooler.Schema
+					specUser = newConnectionPooler.User
+				}
 
-		// in this case also do not forget to install lookup function as for
-		// creating cluster
-		if !oldNeedConnectionPooler || !c.ConnectionPooler.LookupFunction {
-			newConnectionPooler := newSpec.Spec.ConnectionPooler
+				schema := util.Coalesce(
+					specSchema,
+					c.OpConfig.ConnectionPooler.Schema)
 
-			specSchema := ""
-			specUser := ""
+				user := util.Coalesce(
+					specUser,
+					c.OpConfig.ConnectionPooler.User)
 
-			if newConnectionPooler != nil {
-				specSchema = newConnectionPooler.Schema
-				specUser = newConnectionPooler.User
+				if err = lookup(schema, user); err != nil {
+					return NoSync, err
+				}
 			}
 
-			schema := util.Coalesce(
-				specSchema,
-				c.OpConfig.ConnectionPooler.Schema)
-
-			user := util.Coalesce(
-				specUser,
-				c.OpConfig.ConnectionPooler.User)
-
-			if err = lookup(schema, user); err != nil {
-				return NoSync, err
+			if reason, err = c.syncConnectionPoolerWorker(oldSpec, newSpec, role); err != nil {
+				c.logger.Errorf("could not sync connection pooler: %v", err)
+				return reason, err
 			}
 		}
 
-		if reason, err = c.syncConnectionPoolerWorker(oldSpec, newSpec); err != nil {
-			c.logger.Errorf("could not sync connection pooler: %v", err)
-			return reason, err
-		}
-	}
-
-	if oldNeedConnectionPooler && !newNeedConnectionPooler {
-		// delete and cleanup resources
-		if err = c.deleteConnectionPooler(); err != nil {
-			c.logger.Warningf("could not remove connection pooler: %v", err)
+		if oldNeedConnectionPooler && !newNeedConnectionPooler {
+			// delete and cleanup resources
+			otherRole := role
```

Comment:
If I understand this part correctly, this seems to be too detailed and unnecessary for this function. What do you think about introducing a new function defined for `ConnectionPoolerObject` to reset it's value if needed? And then inside one can all present keys.